fair_mango/metrics/base.py

```diff
@@ -236,15 +235,21 @@ def calculate_disparity(
         key = (group_i, group_j)
 
         if method == "difference":
-            result[key] = result_i - result_j
+            yield key, result_i - result_j
         elif method == "ratio":
-            result[key] = result_i / result_j
+            yield key, result_i / result_j
         else:
             raise AttributeError(
-                f"method {method} not recognised. Use 'difference' or "
-                "'ratio' instead."
+                f"method {method} not recognised. Use 'difference' or 'ratio' instead."
             )
 
+
+def calculate_disparity(
+    result_per_groups: np.ndarray, method: Literal["difference", "ratio"]
+) -> dict[tuple, np.ndarray[float]]:
+    result = {}
+    for key, value in disparity_generator(result_per_groups, method):
+        result[key] = value
     return result
```

Comment:
because this function is now super small, you could separate `calculate_difference` from `calculate_ratio`

Reply:
So, `disparity_generator` will be like an initialisation and I should move the actual computation to a separate functions. Which is also possible without a generator.

Is there any particular benefit of using a generator here?

Reply:
yep exactly. Basically, your generator will yield key, result_i, result_j and the higher level function will do the actual computation. Sorry for the lack of explanation

Reply:
I don't think there are any benefits here, that s why I wanted it to be an exercise. It is important to keep generators in mind.

But if you had multiple disparity evaluation functions (difference, ratio, power, mod, absolute, etc.) then it would be the way to go.

Reply:
hmm, from my perspective, if we have multiple operations (difference, ratio, power, mod, absolute, etc.) then a generator wouldn't be the way here too since we are interested in doing the different computations on a small array.

I think we need it when are dealing with a long list of numbers to avoid using too much memory. We can use a generator to process the list by parts.

What do you think?